src/Testing/ZopeTestCase/sandbox.py

```diff
@@ -62,15 +63,17 @@ def app(self):
         return self._app
 
 
-def __bobo_traverse__(self, REQUEST=None, name=None):
-    '''Makes ZPublisher.publish() use the current app object.'''
+@contextlib.contextmanager
+def load_app(module_info):
+    """Let the Publisher use the current app object."""
     app = AppZapper().app()
     if app is not None:
-        return app
-    return self.__old_bobo_traverse__(REQUEST, name)
+        yield app, module_info[1], module_info[2]
+    else:
+        with ZPublisher.WSGIPublisher.__old_load_app__(module_info) as ret:
+            yield ret
 
 
-if not hasattr(ZApplicationWrapper, '__old_bobo_traverse__'):
-    ZApplicationWrapper.__old_bobo_traverse__ = (
-        ZApplicationWrapper.__bobo_traverse__)
-    ZApplicationWrapper.__bobo_traverse__ = __bobo_traverse__
+if not hasattr(ZPublisher.WSGIPublisher, '__old_load_app__'):
+    ZPublisher.WSGIPublisher.__old_load_app__ = ZPublisher.WSGIPublisher.load_app
+    ZPublisher.WSGIPublisher.load_app = load_app
```

Comment:
> Very weird to be monkey-patching a function we are just now adding.

I fully understand your concern. The problem I faced was that `app_wrapper` refers to the global variable `Zope2.bobo_application` retrieved from `ZPublisher.get_module_info()`. So to keep `load_app` untouched one of these have to be monkey patched instead. Lets say I do a monkey patch of `get_module_info`. Then `load_app` closes the database connection too soon and unittests fail. Which means some code needs to be added to mock/hide the call to `app._p_jar.close()` in `load_app()`.

I also tried to patch `Zope2.bobo_application` in `sandbox.Sandboxed._app` and then restore it `sandbox.Sandboxed._close`. With the goal to limit the duration the monkey patch is active.
However some unittests reinitialize `Zope2.bobo_application` after `sandbox.Sandboxed._app` is called and failing again. For example `Products.Five.browser.tests.test_defaultview.test_default_view`.

So my reasoning was: monkey patching `load_app` looks like the lesser evil.